File: school/pos407/wk2/src/Mortgage.java
/*
 * Mortgage.java
 * Week 2 Programming Assignment - Mortgage Payment Calculator
 * POS/407 - Computer Programming II
 * David C. Gibbons, dcgibbons@email.uophx.edu
 *
 * Version | Date       | Description
 * --------|------------|-----------------------------------------------------
 *   1.00  | 2006-01-12 | Initial version.
 */

/**
 * This class performs payment calculations for a mortgage loan.
 */
public class Mortgage {
    private static final int MONTHS_PER_YEAR = 12;

    private double principal;
    private double monthlyRate;
    private int numPayments;
    private double monthlyPayment;

    /**
     * Constructs a new Mortgage object using the provided values.
     *
     * @param principal the starting principal of the mortgage
     * @param rate      the annual percentage rate of the mortgage, e.g. 5.75
     * @param term      the length of the mortgage in years
     */
    public Mortgage(final double principal, final double rate, final int term) {
        setPrincipal(principal);
        setRate(rate);
        setTerm(term);
        calculatePayment();
    }

    /**
     * Sets the principal of the mortgage.
     *
     * @param newPrincipal the new mortgage principal
     */
    public void setPrincipal(final double newPrincipal) {
        principal = newPrincipal;
    }

    /**
     * Sets the annual percentage rate of the mortgage.
     *
     * @param yearlyRate the annual percentage rate, e.g. 5.75
     */
    public void setRate(final double yearlyRate) {
        // determine interest rate per month
        monthlyRate = yearlyRate / 100.0 / MONTHS_PER_YEAR;
    }

    /**
     * Sets the term of the mortgage.
     *
     * @param termInYears the length of the mortgage in years
     */
    public void setTerm(final int termInYears) {
        // determine total number of monthly payments for the given term
        numPayments = termInYears * MONTHS_PER_YEAR;
    }

    public double getPayment() {
        return monthlyPayment;
    }

    /**
     * Calculates the monthly payment of the mortgage.
     */
    protected void calculatePayment() {
        // calculate monthly payment
        monthlyPayment = principal *
                (monthlyRate / (1 - Math.pow(1 + monthlyRate, -numPayments)));
    }
}

File: school/uphoenix/pos402/wk4/MortgageCalculator/MortgageCalculator/Mortgage.vb
' Mortgage.vb
' Week 4 Assignment
' David C. Gibbons
' POS/402 - Visual Basic
' Aron Kuppersmith
' February 5, 2007
'
' Version | Date       | Description
' --------+------------+-----------------------------------------------
'    3.00 | 2007-02-5  | Updated for week 4 assignment.
' --------+------------+-----------------------------------------------
'    2.00 | 2007-01-27 | Updated for week 3 assignment.
' --------+------------+-----------------------------------------------
'    1.00 | 2007-01-20 | Initial version.
' --------+------------+-----------------------------------------------

'
' The Mortgage class provides financial calculations for mortgages.
' 
Public Class Mortgage
    '
    ' Constant for the number of months per year (only to make formulas
    ' easier to read.
    '
    Public Const MONTHS_PER_YEAR = 12

    '
    ' The starting PrincipalAmount of the mortgage.
    '
    Public ReadOnly PrincipalAmount As Decimal

    '
    ' The mortgage interest rate
    '
    Public ReadOnly InterestRate As Double

    '
    ' The MonthlyRate of the mortgage (as opposed to the yearly rate).
    '
    Public ReadOnly MonthlyRate As Double

    '
    ' The Term of the mortgage (years).
    '
    Public ReadOnly Term As Integer

    '
    ' The NumPayments of the mortgate.
    '
    Public ReadOnly NumPayments As Integer

    '
    ' The friendly description of the mortgage terms
    '
    Public ReadOnly Description As String

    '
    ' The MonthlyPayment value contains the currently calculated 
    ' mortgage payment amount.
    '
    Public ReadOnly MonthlyPayment As Decimal

    '
    ' A constructor that requires the specific principal, annual interest rate, 
    ' and mortgage term.
    '
    Public Sub New(ByVal principal As Decimal, ByVal rate As Double, ByVal term As Integer)

        ' save the starting principal amount and mortgage term
        PrincipalAmount = principal
        Me.Term = term
        NumPayments = term * MONTHS_PER_YEAR

        ' calculate the monthly interet rate
        InterestRate = rate
        MonthlyRate = InterestRate / MONTHS_PER_YEAR

        ' build the description
        Description = String.Format("{0:c} @ {1:p} for {2:d} years", _
            PrincipalAmount, InterestRate, term)

        ' calculate the mortgage payment 
        MonthlyPayment = CalculatePayment()

    End Sub

    '
    ' The ToString function provides a friendly description of the Mortgage terms
    '
    Overrides Function ToString() As String
        Return Description
    End Function

    '
    ' Calculates the monthly mortgage payment using a standard financial
    ' calculation.
    '
    Private Function CalculatePayment() As Double
        Dim Payment As Decimal
        Payment = PrincipalAmount * (MonthlyRate / (1 - Math.Pow(1 + MonthlyRate, -NumPayments)))
        Return Payment
    End Function

End Class
